compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt

```diff
@@ -285,28 +289,64 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
   ) {
     companion object {
       fun FunctionDescriptor.toAssistedFactoryFunction(
+        module: ModuleDescriptor,
         factoryClass: KtClassOrObject
       ): AssistedFactoryFunction {
 
+        val containingClass = containingDeclaration as ClassDescriptor
+
         val returnTypeLazy = lazy(NONE) {
-          returnType
-            ?.classDescriptorForType()
-            ?.fqNameSafe
-            ?: throw AnvilCompilationException(
+
+          fun fail(): Nothing {
+            throw AnvilCompilationException(
               "Invalid return type: ${factoryClass.requireFqName()}. An assisted factory's " +
                 "abstract method must return a type with an @AssistedInject-annotated constructor.",
               element = factoryClass
             )
+          }
+
+          // This can be a generic type like `T` if the function is declared in a super type.
+          // Check that it can resolve before using this type.
+          returnType?.classDescriptorOrNull()
+            ?.fqNameSafe
+            ?: factoryClass.resolveGenericKotlinType(
+              module,
+              containingClass.toClassReference(),
+              returnType ?: fail()
+            )
+              ?.fqNameOrNull(module)
+            ?: fail()
         }
         val parameterPairs = lazy(NONE) {
           valueParameters.map { param ->
-            param.name.asString() to param.type.asTypeName()
+
+            val fromDescriptor = param.type
+              // This can be a generic type like `T` if the function is declared in a super type.
+              // Check that it can resolve before using this type.
+              .takeIf { it.classDescriptorOrNull() != null }
+              ?.asTypeName()
+              ?: factoryClass.resolveGenericKotlinType(
+                module,
+                containingClass.toClassReference(),
+                param.type
+              )
+                ?.requireTypeName(module)
```

Comment:
I find the formatting a little weird here. I think it's easier to read if `?.requireTypeName()` is on the previous line.
```suggestion
              )?.requireTypeName(module)
```